### ISSUE TITLE
Implement client-side buffering for publishing messages

### ISSUE THREAD

**Issue by garrettjonesgoogle** [GoogleCloudPlatform/google-cloud-java#384]
This functionality needs to support these settings:
- max batched messages
- max batched bytes
- max batching delay

**Comment by aozarov** [comment]
When doing so we should consider that gRPC supports client-side streaming (so gax would need to support it) and that pubsub is planning to add support for it as well for publishing messages.

**Comment by garrettjonesgoogle** [comment]
I think it would be best to support the streaming as a follow-on feature.

**Comment by piotrpolatowski** [comment]
Have you also thought about buffered version of `MessageConsumer pullAsync(String subscription, MessageProcessor callback, PullOption... options)`?

Right now I've done it utilizing `Future<Iterator<ReceivedMessage>> pullAsync(String subscription, int maxMessages)` however I have to re-implement Ack Deadline Renewing and create a separate buffer from internal `PubSub` queue.

Moreover, what about making `AckDeadlineRenewer` class public? One unit tested class to re-implement less.

**Comment by mziccard** [comment]
> Have you also thought about buffered version of MessageConsumer pullAsync(String subscription, MessageProcessor callback, PullOption... options)?

Can you clarify what you mean with this? Right now you can specify how many pending callbacks you want by providing the [`maxQueuedCallbacks` option](https://github.com/GoogleCloudPlatform/gcloud-java/blob/master/gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/PubSub.java#L117). And you can also control the amount of callbacks being executed in parallel by passing your own executor with the [`executorFactory` option](https://github.com/GoogleCloudPlatform/gcloud-java/blob/master/gcloud-java-pubsub/src/main/java/com/google/cloud/pubsub/PubSub.java#L136).

**Comment by piotrpolatowski** [comment]
What if I want to write messages to a storage in batches, not one by one.
I'd like to acknowledge messages only when the batch was successfully stored, `maxQueuedCallbacks` doesn't solve that problem, right?

**Comment by mziccard** [comment]
> What if I want to write messages to a storage in batches, not one by one?

I see what you mean. This sounds like a kind of batching you need to do in your application code and does not fit well at our library layer. However, what I believe could help your case might be a `Future<Iterator<ReceivedMessage>> pullAsync(String subscription, int maxMessages)` method that renews message deadlines until they are explicitly acked/nacked (and not just consumed). This way you could pull messages, buffer them, batch-store them and ack the batch if store was successfull, without caring about renewals. Does this make sense?

**Comment by piotrpolatowski** [comment]
Yeah, this make sense. Does it deserve a new issue?

**Comment by garrettjonesgoogle** [comment]
The feature as originally described has been implemented for a while now in the spi layer, but not in the top-most layer. We should look at what it would take to add it there too.

**Comment by garrettjonesgoogle** [comment]
@pongad is working on a new version of pubsub which will enable this.

**Comment by garrettjonesgoogle** [comment]
The Pub/Sub rewrite has been merged and released. The `Publisher` class provides this functionality. Closing out this issue.
